jfix-stdlib-id-generator/src/main/kotlin/ru/fix/stdlib/id/generator/SynchronizedIdGenerator.kt

```diff
@@ -0,0 +1,53 @@
+package ru.fix.stdlib.id.generator
+
+import java.time.Clock
+
+class SynchronizedIdGenerator(
+        private val bitsConfiguration: BitsConfiguration,
+        private val startOfTime: Long,
+        serverId: Long,
+        private val clock: Clock,
+        counter: Long = 0L
+) : IdGenerator {
+
+    private val serverIdPart: Long
+    private var counter: Long
+    private var idTime: Long = 0L
+
+    init {
+        require(serverId <= bitsConfiguration.serverPartMask) {
+            "serverId cannot be greater than ${bitsConfiguration.serverPartMask}"
+        }
+
+        require(startOfTime < clock.millis()) {
+            "startOfTime cannot be greater than $startOfTime"
+        }
+
+        this.serverIdPart = serverId and bitsConfiguration.serverPartMask
+        this.counter = counter
+    }
+
+    @Synchronized
+    override fun nextId(): Long {
+        counter++
+        return if (counterIsOverflow(counter)) {
+            idTime++
+            counter = 0
+            generateId(counter)
+        } else {
+            generateId(counter)
+        }
+    }
+
+    private fun generateId(counterValue: Long): Long {
+        catchUpIdTimeWithCurrentTimeIfNeeded()
+        val tsPart = ((idTime - startOfTime) and bitsConfiguration.timePartMask) shl bitsConfiguration.serverPartBits + bitsConfiguration.counterPartBits
```

Comment:
shl and plus in kotlin, not sure about operator precedence: it will be better to explicitly use braces here